staging/src/k8s.io/pod-security-admission/cmd/webhook/app/server.go

```diff
@@ -0,0 +1,261 @@
+		err = fmt.Errorf("contentType=%s, expected application/json", contentType)
+		klog.ErrorS(err, "unable to process a request with an unknown content type", "content type", contentType)
+		writeAdmissionResponse(w, admission.BadRequestResponse("unable to process a request with a non-json content type"))
+		return
+	}
+
+	// Both v1 and v1beta1 AdmissionReview types are exactly the same, so the v1beta1 type can
+	// be decoded into the v1 type. However the runtime codec's decoder guesses which type to
+	// decode into by type name if an Object's TypeMeta isn't set. By setting TypeMeta of an
+	// unregistered type to the v1 GVK, the decoder will coerce a v1beta1 AdmissionReview to v1.
+	// The actual AdmissionReview GVK will be used to write a typed response in case the
+	// webhook config permits multiple versions, otherwise this response will fail.
+	review := &admissionv1.AdmissionReview{}
+	review.SetGroupVersionKind(admissionv1.SchemeGroupVersion.WithKind("AdmissionReview"))
+	if _, _, err := codecs.UniversalDeserializer().Decode(body, nil, review); err != nil {
+		klog.ErrorS(err, "unable to decode the request")
+		writeAdmissionResponse(w, admission.BadRequestResponse("unable to decode the request"))
+		return
+	}
+	klog.V(1).InfoS("received request", "UID", review.Request.UID, "kind", review.Request.Kind, "resource", review.Request.Resource)
+
+	attributes := admission.RequestAttributes(review.Request, s)
+	response := s.delegate.Validate(ctx, attributes)
+	review.Response = response
+	writeResponse(w, review)
+}
+
+// Server implements the admission.Decoder interface.
+func (s *Server) DecodeRaw(rawObj runtime.RawExtension, into runtime.Object) error {
+	if len(rawObj.Raw) == 0 {
+		return fmt.Errorf("empty object")
+	}
+	deserializer := codecs.UniversalDeserializer()
+	return runtime.DecodeInto(deserializer, rawObj.Raw, into)
+}
+
+// Setup creates an Admission object to handle the admission logic.
+func Setup(opts *options.Options) (*Server, error) {
+	if errs := opts.Validate(); len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+
+	s := &Server{}
+	opts.SecureServing.ApplyTo(&s.secureServing)
+	opts.InsecureServing.ApplyTo(&s.insecureServing)
+
+	if s.secureServing == nil && s.insecureServing == nil {
+		return nil, errors.New("no serving options configured")
+	}
+
+	// Load Kube Client
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", opts.KubeConfigFile)
+	if err != nil {
+		return nil, err
+	}
+	restclient.AddUserAgent(kubeConfig, "podsecurity-webhook")
+	client, err := clientset.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	s.informerFactory = kubeinformers.NewSharedInformerFactory(client, time.Second*30)
+	namespaceInformer := s.informerFactory.Core().V1().Namespaces()
+	namespaceLister := namespaceInformer.Lister()
+	podLister := s.informerFactory.Core().V1().Pods().Lister() // FIXME: consider switching to just use the client
```

Comment:
you thinking the memory/network cost of watching and holding all pod events is worse than the cost of listing all pods in a namespace on edit of namespace policy labels? I'd tend to agree, since the latter is rare

Reply:
Yeah, exactly. I meant to fix this before sending it out. Will do.